Sprawozdanie 2

Łączenie z bazą danych i instrukcja select

In [1]:
from sqlalchemy import create_engine
import pandas as pd

#Łączenie się z bazą danych
db_string = "postgresql://wbauer_adb:adb2020@pgsql-196447.vipserv.org:5432/wbauer_adb"

db = create_engine(db_string)

connection_sqlalchemy = db.connect()

Zadanie 1
    
Ile kategorii filmów mamy w wypożyczalni?


In [11]:
df = pd.read_sql('select count(name) from category', con=connection_sqlalchemy)
df

count
0     16

Odpowiedź: W wypożyczalni jest 16 kategorii filmów.

Zadanie 2

Wyświetl listę kategorii w kolejności alfabetycznej.

In [3]:
df = pd.read_sql('select * from category order by name', con=connection_sqlalchemy)
df

category_id         name         last_update
0             1       Action 2006-02-15 09:46:27
1             2    Animation 2006-02-15 09:46:27
2             3     Children 2006-02-15 09:46:27
3             4     Classics 2006-02-15 09:46:27
4             5       Comedy 2006-02-15 09:46:27
5             6  Documentary 2006-02-15 09:46:27
6             7        Drama 2006-02-15 09:46:27
7             8       Family 2006-02-15 09:46:27
8             9      Foreign 2006-02-15 09:46:27
9            10        Games 2006-02-15 09:46:27
10           11       Horror 2006-02-15 09:46:27
11           12        Music 2006-02-15 09:46:27
12           13          New 2006-02-15 09:46:27
13           14       Sci-Fi 2006-02-15 09:46:27
14           15       Sports 2006-02-15 09:46:27
15           16       Travel 2006-02-15 09:46:27

Zadanie 3

Znajdź najstarszy i najmłodszy film do wypożyczenia.

In [5]:
df = pd.read_sql('select * from film order by release_year', con=connection_sqlalchemy)
df

film_id              title  \
0        133    Chamber Italian   
1        384   Grosse Wonderful   
2          8    Airport Pollock   
3         98  Bright Encounters   
4          1   Academy Dinosaur   
..       ...                ...   
995      996     Young Language   
996      997         Youth Kick   
997      998       Zhivago Core   
998      999  Zoolander Fiction   
999     1000          Zorro Ark   

                                           description  release_year  \
0    A Fateful Reflection of a Moose And a Husband ...          2006   
1    A Epic Drama of a Cat And a Explorer who must ...          2006   
2    A Epic Tale of a Moose And a Girl who must Con...          2006   
3    A Fateful Yarn of a Lumberjack And a Feminist ...          2006   
4    A Epic Drama of a Feminist And a Mad Scientist...          2006   
..                                                 ...           ...   
995  A Unbelieveable Yarn of a Boat And a Database ...          2006   
996  A Touching Drama of a Teacher And a Cat who mu...          2006   
997  A Fateful Yarn of a Composer And a Man who mus...          2006   
998  A Fateful Reflection of a Waitress And a Boat ...          2006   
999  A Intrepid Panorama of a Mad Scientist And a B...          2006   

     language_id  rental_duration  rental_rate  length  replacement_cost  \
0              1                7         4.99     117             14.99   
1              1                5         4.99      49             19.99   
2              1                6         4.99      54             15.99   
3              1                4         4.99      73             12.99   
4              1                6         0.99      86             20.99   
..           ...              ...          ...     ...               ...   
995            1                6         0.99     183              9.99   
996            1                4         0.99     179             14.99   
997            1                6         0.99     105             10.99   
998            1                5         2.99     101             28.99   
999            1                3         4.99      50             18.99   

    rating             last_update  \
0    NC-17 2013-05-26 14:50:58.951   
1        R 2013-05-26 14:50:58.951   
2        R 2013-05-26 14:50:58.951   
3    PG-13 2013-05-26 14:50:58.951   
4       PG 2013-05-26 14:50:58.951   
..     ...                     ...   
995      G 2013-05-26 14:50:58.951   
996  NC-17 2013-05-26 14:50:58.951   
997  NC-17 2013-05-26 14:50:58.951   
998      R 2013-05-26 14:50:58.951   
999  NC-17 2013-05-26 14:50:58.951   

                                special_features  \
0                                     [Trailers]   
1                            [Behind the Scenes]   
2                                     [Trailers]   
3                                     [Trailers]   
4            [Deleted Scenes, Behind the Scenes]   
..                                           ...   
995                [Trailers, Behind the Scenes]   
996                [Trailers, Behind the Scenes]   
997                             [Deleted Scenes]   
998                   [Trailers, Deleted Scenes]   
999  [Trailers, Commentaries, Behind the Scenes]   

                                              fulltext  
0    'chamber':1 'fate':4 'husband':11 'italian':2 ...  
1    'australia':18 'cat':8 'drama':5 'epic':4 'exp...  
2    'airport':1 'ancient':18 'confront':14 'epic':...  
3    'boat':20 'bright':1 'conquer':14 'encount':2 ...  
4    'academi':1 'battl':15 'canadian':20 'dinosaur...  
..                                                 ...  
995  'administr':12 'boat':8 'boy':17 'databas':11 ...  
996  'boat':22 'cat':11 'challeng':14 'drama':5 'ki...  
997  'boy':16 'canadian':19 'compos':8 'core':2 'fa...  
998  'ancient':19 'boat':11 'china':20 'discov':14 ...  
999  'ark':2 'boy':12,17 'intrepid':4 'mad':8 'mona...  

[1000 rows x 13 columns]

Odpowiedź: Wszystkie filmy są z roku 2006.

Zadanie 4

Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?

In [9]:
df = pd.read_sql("select count(rental_date) from rental where rental_date BETWEEN '2005-07-01 00:00:00'::timestamp AND '2005-08-01 00:00:00'::timestamp",con=connection_sqlalchemy)
df

count
0   6709

Zadanie 5


Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?

In [10]:
df = pd.read_sql("select count(rental_date) from rental where rental_date BETWEEN '2010-01-01 00:00:00'::timestamp AND '2011-02-01 00:00:00'::timestamp",con=connection_sqlalchemy)
df

count
0      0

Zadanie 6

Znajdź największą płatność wypożyczenia.

In [13]:
df = pd.read_sql("select amount from payment order by amount",con=connection_sqlalchemy)
df

amount
0        0.00
1        0.00
2        0.00
3        0.00
4        0.00
...       ...
14591   11.99
14592   11.99
14593   11.99
14594   11.99
14595   11.99

[14596 rows x 1 columns]

Odpowiedź: Największa płatność wypoyczenia wynosi 11.99 zł.

Zadanie 7

Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.

In [25]:
df = pd.read_sql_query("select * from customer_list where country = 'Poland' or country = 'Nigeria' or country = 'Bangladesh'",con=connection_sqlalchemy)
df

id              name                         address zip code  \
0   284     Sonia Gregory              1279 Udine Parkway    75860   
1   318       Brian Wyman                 1769 Iwaki Lane    25787   
2   572   Sidney Burleson             1405 Hagonoy Avenue    86587   
3   409    Rodney Moeller          1936 Lapu-Lapu Parkway     7122   
4   332    Stephen Qualls                1838 Tabriz Lane     1195   
5   286       Velma Lucas            1155 Liaocheng Place    22650   
6    80      Marilyn Ross                1888 Kabul Drive    20936   
7   198      Elsie Kelley                 955 Bamenda Way     1545   
8   103   Gladys Hamilton                 1177 Jelets Way     3305   
9   329    Frank Waggoner             1816 Bydgoszcz Loop    64308   
10  128   Marjorie Tucker        1050 Garden Grove Avenue     4999   
11   18      Carol Garcia                320 Brest Avenue    43331   
12  270       Leah Curtis                  906 Goinia Way    83565   
13  501       Ruben Geary          414 Mandaluyong Street    16370   
14  277      Olga Jimenez                556 Baybay Manor    55802   
15  156   Bertha Ferguson               1367 Yantai Manor    21294   
16  280    Tracey Barrett  1006 Santa Brbara dOeste Manor    36229   
17  250         Jo Fowler            786 Stara Zagora Way    98332   
18  571  Johnnie Chisholm      1501 Pangkal Pinang Avenue      943   
19  562     Wallace Slone                 1342 Sharja Way    93655   
20   21    Michelle Clark  262 A Corua (La Corua) Parkway    34418   
21  380   Russell Brinson     1642 Charlotte Amalie Drive    75442   
22  558  Jimmie Eggleston            505 Madiun Boulevard    97271   
23  232    Constance Reid               885 Yingkou Manor    31390   

           phone             city     country   notes  sid  
0   195003555232       Benin City     Nigeria  active    1  
1   556100547674        Bydgoszcz      Poland  active    1  
2   867287719310      Czestochowa      Poland  active    1  
3   653436985797        Deba Habe     Nigeria  active    2  
4    38988715447            Dhaka  Bangladesh  active    1  
5   558236142492     Effon-Alaiye     Nigeria  active    1  
6   701457319790              Ife     Nigeria  active    1  
7   768481779568           Ikerre     Nigeria  active    2  
8   484292626944           Ilorin     Nigeria  active    1  
9   965273813662         Jamalpur  Bangladesh  active    2  
10  973047364353  Jastrzebie-Zdrj      Poland  active    1  
11  747791594069           Kaduna     Nigeria  active    2  
12  701767622697           Kalisz      Poland  active    1  
13   52709222667           Lublin      Poland  active    1  
14  363982224739        Ogbomosho     Nigeria  active    2  
15  889538496300             Ondo     Nigeria  active    1  
16   85059738746              Owo     Nigeria  active    2  
17  716256596301              Oyo     Nigeria  active    2  
18  770864062795            Plock      Poland  active    2  
19  946114054231           Sokoto     Nigeria  active    1  
20  892775750063          Tangail  Bangladesh  active    1  
21  821476736117            Tychy      Poland  active    1  
22  970638808606          Wroclaw      Poland  active    1  
23  588964509072            Zaria     Nigeria  active    2

Zadanie 8

Gdzie mieszkają członkowie personelu?

In [31]:
df = pd.read_sql_query("select address, city, country from staff_list ",con=connection_sqlalchemy)
df

address        city    country
0     23 Workhaven Lane  Lethbridge     Canada
1  1411 Lillydale Drive   Woodridge  Australia

Zadanie 9

Ilu pracowników mieszka w Argentynie lub Hiszpanii?

In [32]:
df = pd.read_sql_query("select count(country) from staff_list where country = 'Argentine' or country = 'Spain'",con=connection_sqlalchemy)
df

count
0      0

Zadanie 10

Jakie kategorie filmów zostały wypożyczone przez klientów?

In [48]:
df = pd.read_sql('select category from sales_by_film_category where total_sales > 0', con=connection_sqlalchemy)
df

category
0        Sports
1        Sci-Fi
2     Animation
3         Drama
4        Comedy
5           New
6        Action
7       Foreign
8         Games
9        Family
10  Documentary
11       Horror
12     Classics
13     Children
14       Travel
15        Music

Zadanie 11

Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.

In [62]:
df = pd.read_sql("select distinct category from film_list where fid in (select film_id from inventory where inventory_id in (select inventory_id from rental where customer_id in (select id from customer_list where country = 'United States')))", con=connection_sqlalchemy)
df

category
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

Zadanie 12

Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presley

In [87]:
df = pd.read_sql("select title from film where film_id in(select film_id from film_actor where actor_id in (select actor_id from actor where (first_name like 'Olympia' and last_name like 'Pfeiffer') or (first_name like 'Julia' and last_name like 'Zellweger') or (first_name like 'Ellen' and last_name like 'Presley')))", con=connection_sqlalchemy)
df

title
0      Express Lonely
1     Bilko Anonymous
2    Effect Gladiator
3   Contact Anonymous
4         Others Soup
..                ...
64      Hanky October
65    Tourist Pelican
66   Detective Vision
67       None Spiking
68     Homeward Cider

[69 rows x 1 columns]